### DISCUSSION TITLE
some ideas for add-ones to transfer dwg to dxf and print cad

### DISCUSSION THREAD

**Discussion by chibai** [mozman/ezdxf#724] [Ideas]
We can actually make full use of AutoCad accoreconsole.exe to do something that is original?authentic？
Here's the code if you can get it~

code transfer dwg to dxf with computer installed AutoCad
```python
def transfer_dwg2dxf_if_necessary(input_path: str) -> None:
    if input_path.endswith('.dwg'):
        input_dxf_path = input_path.replace('.dwg', '.dxf')
        if os.path.exists(input_dxf_path):
            os.remove(input_dxf_path)
        # create script save dwg file to dxf file by accoreconsole
        lock.acquire()
        with open(save2dxf_script_path, 'w') as f:
            f.write('AUDIT\n')
            f.write('Y\n')
            f.write('SAVE ' + input_dxf_path + '\n')
            f.write('16\n')
            f.write('QUIT\n')
        #cmd_line = 'accoreconsole.exe /i "' + input_path + '" /s "' + save2dxf_script_path + '"'
        s = time.time()
        result = subprocess.run(["accoreconsole.exe", "/i", input_path, "/s", save2dxf_script_path], stdout=subprocess.PIPE)
        e = time.time()
        lock.release()
        if result.returncode  != 0:
            print('fail to transfer from dwg 2 dxf')
            return False
        print("transfer time", e - s)
        return input_dxf_path
    else:
        return input_path
```

code print CAD file with computer installed AutoCad, while '2048SQUARE' is predefined papersize
```python
def pt2str(pt: Tuple[float, float]) -> str:
    return str(round(pt[0], 2)) + ',' + str(round(pt[1], 2))

def print_png(input_path, output_path: str, ext: BoundingBox) -> None:
    lock.acquire()
    with open(cad2png_script_path, 'w') as f:
        # Command:
        f.write('FILEDIA\n')
        # Enter new value for FILEDIA <1>:
        f.write('0\n')
        # USC command
        f.write('UCS\n')
        # Word coordinate
        f.write('W\n')
        # Command:
        f.write('-PLOT\n')
        # Detailed plot configuration? [Yes/No] <No>:
        f.write('Y\n')
        # Enter a layout name or [?] <Model>:
        f.write('Model\n')
        # Enter an output device name or [?] <None>:
        f.write('PublishToWeb PNG.pc3\n')
        # Enter paper size or [?] <ANSI A (11.00 x 8.50 Inches)>
        f.write('2048SQUARE\n')
        # Enter drawing orientation [Portrait/Landscape] <Portrait>
        f.write('P\n')
        # Plot upside down? [Yes/No] <No>
        f.write('N\n')
        # Enter plot area [Display/Extents/Limits/View/Window] <Display>
        f.write('W\n')
        # 输入窗口的左下角 <0.000000,0.000000>
        f.write(pt2str(ext.extmin) + '\n')
        # 输入窗口的右上角 <0.000000,0.000000>
        f.write(pt2str(ext.extmax) + '\n')
        # Enter plot scale (Plotted Inches=Drawing Units) or [Fit] <Fit>
        f.write('F\n')
        # Enter plot offset (x,y) or [Center] <0.00,0.00>
        f.write('C\n')
        # Plot with plot styles? [Yes/No] <Yes>:
        f.write('Y\n')
        # Enter plot style table name or [?] (enter . for none) <>:
        f.write('.\n')
        # Plot with lineweights? [Yes/No] <Yes>
        f.write('Y\n')
        # Enter shade plot setting [As displayed/legacy Wireframe/legacy Hidden/Visualstyles/Rendered] <As displayed>:
        f.write('A\n')
        # Enter file name <>:
        f.write('"' + output_path + '"\n')
        # Save changes to page setup? [Yes/No]
        f.write('N\n')
        # Proceed with plot [Yes/No] <Y>
        f.write('Y\n')
        # quit the command
        f.write('QUIT\n')
        f.write('Y\n')
    s = time.time()
    #cmd_line = 'accoreconsole.exe /i "' + input_path + '" /s "' + cad2png_script_path + '"'
    result = subprocess.run(["accoreconsole.exe", "/i", input_path, "/s", cad2png_script_path], stdout=subprocess.PIPE)
    e = time.time()
    lock.release()
    if result.returncode  != 0:
        print('fail to print image')
    print('print image: ', e - s)
```

These codes are actually running in our startup company web software online, but our starup failed, so I come to share the idea.
We can actually make full use AutoCad accoreconsole.exe to do something

**Comment by mozman** [comment]
I googled "accoreconsole.exe" and found that it also ships with DWG TrueView 2019 and later. So, at least for Windows users, it is possible to get this application for free. I will consider a similar addon as for ODAFC if it works on my machine.

**Comment by mozman** [comment]
I tested this simple script to convert a DWG to a DXF:
```text
AUDIT
Y
SAVE "C:\Users\mozman\Desktop\Now\ezdxf\acconsole\output.dxf"
16
QUIT
````
Execution by accoreconsole.exe , where`<fullpath>` is a placeholder to shorten the text:
`accoreconsole.exe /i "<fullpath>\input.dwg" /s "<fullpath>\script.scr"`

and got this result:
```text
Redirect stdout (file: C:\Users\mozman\AppData\Local\Temp\accc152842).
AcCoreConsole: StdOutConsoleMode: processed-output: enabled,auto
AutoCAD Core Engine Console - Copyright 2021 Autodesk, Inc.  All rights reserved. (S.154.0.0)

Execution Path:
C:\Program Files\Autodesk\DWG TrueView 2022 - English\accoreconsole.exe
Current Directory: C:\Users\mozman\Desktop\Now\ezdxf\acconsole

Version Number: S.154.0.0 (UNICODE)
LogFilePath has been set to the working folder.
Loading AEC Base...
Loading AEC Base Extended...
Loading AEC Core...
Loading AEC Project Base...
Loading AEC Architectural Base...
Loading AEC Schedule...
Regenerating layout.
Regenerating model.
Press ENTER to continue:
**** System Variable Changed ****
1 of the monitored system variables has changed from the preferred value. Use SYSVARMONITOR command to view changes.



Enter BACKSPACE to interrupt script.



AUDIT
Unknown command "AUDIT".  Press F1 for help.
```

It doesn't also recognize the commands `SAVE` or `HELP`. I assume a full AutoCAD license is required to execute scripts.

**Reply by chibai** [reply]
That's wired, Here's some result for my AutoCad accoreconsole
accoreconsole.exe -help
```
C:\Users\chibai>accoreconsole.exe -help
Redirect stdout (file: C:\Users\chibai\AppData\Local\Temp\accc67803).
AcCoreConsole: StdOutConsoleMode: processed-output: enabled,auto
AutoCAD Core Engine Console - Copyright 2019 Autodesk, Inc.  All rights reserved. (Q.47.0.0)

Execution Path:
D:\Autodesk\AutoCAD 2020\accoreconsole.exe

Version Number: Q.47.0.0 (UNICODE)

Usage:
AcCoreConsole.exe [/i <input dwg>] /s <script>[/product <product>] [/l <language>] [/isolate <userid> <userDataFolder>] [/readonly] [/p[rofile] <profile>]

Example:
AcCoreConsole.exe /i 8th_floor.dwg /s test.scr /l en-US
正在重生成模型。

**** 系统变量更改 ****
已从首选值更改 3 个监视系统变量。使用 SYSVARMONITOR 命令以查看更改。
命令:
命令:
命令:
命令:
```

**Comment by chibai** [comment]
try "accoreconsole.exe /i "<fullpath>\input.dwg" 
And try AUDIT or SAVE within accoreconsole CMD

**Comment by mozman** [comment]
The console does not accept any input, just beeps...

![image](https://user-images.githubusercontent.com/4230966/183345035-2f40b889-b7f5-460a-8c33-3517e0e23f90.png)

I assume the `accoreconsole.exe` just opens the console of the application and passes the script and the DWG TrueView console does not accept any input, it just shows some status messages at the console.

**Reply by chibai** [reply]
OK, my idea only work for AutoCad Licensen then..

**Comment by mozman** [comment]
I will not add an add-on to wrap the `accoreconsole` despite the fact that this add-on would be similar to the ODAFC add-on, but there are some differences:

- The ODA FileConverter is accessible to everyone for free of Windows, macOS and Linux.
- I can maintain the ODAFC add-on, because I have access to the application.
- The ODAFC has an impractical command line interface that is hidden by the add-on, that was the main reason to include this add-on in `ezdxf`. There is only a thematic relationship between the ODAFC and `ezdxf`, but the add-on could also be a standalone Python package.
- Your script shows that the usage of the `accoreconsole` is straight forward for experienced AutoCAD users, and if they can use a search-engine they may find this discussion thread 😄

**Reply by chibai** [reply]
Agree~!

**Reply by changkeke** [reply]
> I will not add an add-on to wrap the `accoreconsole` despite the fact that this add-on would be similar to the ODAFC add-on, but there are some differences:
> 
> * The ODA FileConverter is accessible to everyone for free of Windows, macOS and Linux.
> * I can maintain the ODAFC add-on, because I have access to the application.
> * The ODAFC has an impractical command line interface that is hidden by the add-on, that was the main reason to include this add-on in `ezdxf`. There is only a thematic relationship between the ODAFC and `ezdxf`, but the add-on could also be a standalone Python package.
> * Your script shows that the usage of the `accoreconsole` is straight forward for experienced AutoCAD users, and if they can use a search-engine they may find this discussion thread 😄



> I will not add an add-on to wrap the `accoreconsole` despite the fact that this add-on would be similar to the ODAFC add-on, but there are some differences:
> 
> * The ODA FileConverter is accessible to everyone for free of Windows, macOS and Linux.
> * I can maintain the ODAFC add-on, because I have access to the application.
> * The ODAFC has an impractical command line interface that is hidden by the add-on, that was the main reason to include this add-on in `ezdxf`. There is only a thematic relationship between the ODAFC and `ezdxf`, but the add-on could also be a standalone Python package.
> * Your script shows that the usage of the `accoreconsole` is straight forward for experienced AutoCAD users, and if they can use a search-engine they may find this discussion thread 😄

I saw the ODA official website. The ODA FileConverter is 'Free Trail'?  I'm not sure it is free or not?

**Reply by mozman** [reply]
The converter is free, the trail is for the ODA membership. Just click on "Try for Free".

**Comment by changkeke** [comment]
that's cool!thanks!




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "mozman/ezdxf"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年5月19日(星期五) 下午4:27
***@***.***&gt;;
抄送:&nbsp;"Bounty ***@***.******@***.***&gt;;
主题:&nbsp;Re: [mozman/ezdxf] some ideas for add-ones to transfer dwg to dxf and print cad (Discussion #724)





 
The converter is free, the trail is for the ODA membership. Just click on "Try for Free".
